### PR TITLE
feat: concurrent da_dispatcher

### PR DESCRIPTION
**Pull request by juan518munoz**
## Checklist

- [ ] PR title corresponds to the body of PR (we generate changelog entries from PRs).
- [ ] Tests for the changes have been added / updated.
- [ ] Documentation comments have been added / updated.
- [ ] Code has been formatted via `zk fmt` and `zk lint`.

### CODE REVIEW

**Review by gianbelinche — APPROVED** [review]
LGTM

File: core/node/da_dispatcher/src/da_dispatcher.rs [hunk 1]
```diff
@@ -14,11 +20,14 @@ use zksync_types::L1BatchNumber;
 
 use crate::metrics::METRICS;
 
+const NO_FIRST_BATCH: i64 = -1;
```
Comment (juanbono) [inline core/node/da_dispatcher/src/da_dispatcher.rs:23]: I don't like having this constant here. We should rethink a bit how we do this in the future.

File: core/node/da_dispatcher/src/da_dispatcher.rs [hunk 2]
```diff
@@ -27,178 +36,296 @@ impl DataAvailabilityDispatcher {
         config: DADispatcherConfig,
         client: Box<dyn DataAvailabilityClient>,
     ) -> Self {
+        let request_semaphore = Arc::new(tokio::sync::Semaphore::new(100));
```
Comment (juanbono) [inline core/node/da_dispatcher/src/da_dispatcher.rs:39]: the value passed to the Semaphore constructor can be moved to a constant and documented .

File: core/node/da_dispatcher/src/da_dispatcher.rs [hunk 3]
```diff
@@ -27,178 +36,296 @@ impl DataAvailabilityDispatcher {
...
+        tokio::select! {
+            _ = subtasks => {},
+        }
+        Ok(())
+    }
+
+    async fn dispatch_batches(&self, stop_receiver: Receiver<bool>) -> anyhow::Result<()> {
+        let (tx, mut rx) = mpsc::channel(100);
+
+        let next_expected_batch = Arc::new(AtomicI64::new(NO_FIRST_BATCH));
```
Comment (juanbono) [inline core/node/da_dispatcher/src/da_dispatcher.rs:71]: can you add some docs here about why this atomic is needed?

File: core/node/da_dispatcher/src/da_dispatcher.rs [hunk 4]
```diff
@@ -27,178 +36,296 @@ impl DataAvailabilityDispatcher {
...
+
+        let stop_receiver_clone = stop_receiver.clone();
+        let pool_clone = self.pool.clone();
+        let config_clone = self.config.clone();
+        let next_expected_batch_clone = next_expected_batch.clone();
+        let pending_blobs_reader = tokio::spawn(async move {
+            // Used to avoid sending the same batch multiple times
+            let mut pending_batches = HashSet::new();
+            // let pair = cvar_pair_clone.clone();
```
Comment (juanbono) [inline core/node/da_dispatcher/src/da_dispatcher.rs:80]: remove this commented line if it is not needed.

**Review by juanbono — CHANGES_REQUESTED** [review]
See comments

File: core/node/da_dispatcher/src/da_dispatcher.rs [hunk 5]
```diff
@@ -27,178 +36,296 @@ impl DataAvailabilityDispatcher {
...
+                    Ok::<(), anyhow::Error>(())
+                });
+                spawned_requests.push(request);
             }
-        }
+            join_all(spawned_requests).await;
+            Ok::<(), anyhow::Error>(())
+        });
 
-        tracing::info!("Stop signal received, da_dispatcher is shutting down");
```
[the remark this answers is not in the document]
Comment (juan518munoz) [reply core/node/da_dispatcher/src/da_dispatcher.rs:77]: Mistake, added it again.

File: core/node/da_dispatcher/src/da_dispatcher.rs (hunk 4 above)
Comment (juan518munoz) [reply core/node/da_dispatcher/src/da_dispatcher.rs:81]: Done!

File: core/node/da_dispatcher/src/da_dispatcher.rs (hunk 3 above)
Comment (juan518munoz) [reply core/node/da_dispatcher/src/da_dispatcher.rs]: I've replaced the `AtomicI64` in favour of a `Mutex` that holds an `Option`, which makes the code clearer. I can add some docs if you still consider it necessary.

File: core/node/da_dispatcher/src/da_dispatcher.rs (hunk 2 above)
Comment (juan518munoz) [reply core/node/da_dispatcher/src/da_dispatcher.rs]: Same as above, it's now located in the config file.

File: core/node/da_dispatcher/src/da_dispatcher.rs (hunk 1 above)
Comment (juan518munoz) [reply core/node/da_dispatcher/src/da_dispatcher.rs]: It has been moved to the corresponding config file.

**Review by juanbono — APPROVED** [review]

**Pull request merged**